clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -670,29 +674,63 @@ int nextTimeoutMs() {
+         */
+        private ApiException metadataAuthException = null;
+
+        /**
+         * Check if the AdminClient metadata is usable.
          *
-         * @param prevMetadataVersion       The previous metadata version which wasn't usable.
-         * @return                          null if the metadata is usable; the current metadata
-         *                                  version otherwise
+         * @return                  null, if the metadata is usable.
+         *                          StaleMetadataException, if the metadata is not current.
+         *                          AuthenticationException, if there was an authentication issue
+         *                          while fetching the metadata.
          */
-        private Integer checkMetadataReady(Integer prevMetadataVersion) {
-            if (prevMetadataVersion != null) {
-                if (prevMetadataVersion == metadata.version())
-                    return prevMetadataVersion;
+        private ApiException checkMetadata() {
+            // If someone has requested a metadata update, or if the current metadata is
+            // just the bootstrap cluster, request a new metadata update.
+            if (shouldRequestMetadata.getAndSet(false) ||
+                    metadata.fetch().isBootstrapConfigured()) {
+                if (lowestValidMetadataVersion == lastSeenMetadataVersion) {
+                    lastSeenMetadataVersion = metadata.requestUpdate();
+                    lowestValidMetadataVersion = lastSeenMetadataVersion + 1;
+                    log.debug("Requesting new metadata version {}.", lowestValidMetadataVersion);
+                }
             }
-            Cluster cluster = metadata.fetch();
-            if (cluster.nodes().isEmpty()) {
-                log.trace("Metadata is not ready yet. No cluster nodes found.");
-                return metadata.requestUpdate();
+            int version = metadata.version();
+            if (version > lastSeenMetadataVersion) {
+                metadataAuthException = metadata.getAndClearAuthenticationException();
+                if (metadataAuthException != null) {
+                    log.warn("Metadata update {} had a metadata authentication exception", version,
+                        metadataAuthException);
+                } else {
+                    log.debug("Received new metadata version {}.", version);
+                }
+                lastSeenMetadataVersion = version;
             }
-            if (cluster.controller() == null) {
-                log.trace("Metadata is not ready yet. No controller found.");
-                return metadata.requestUpdate();
+            if (version < lowestValidMetadataVersion) {
+                return INVALID_METADATA_EXCEPTION;
             }
-            if (prevMetadataVersion != null) {
-                log.trace("Metadata is now ready.");
+            if (metadataAuthException != null) {
```

Comment:
Should this if block be placed above the if block on line 729 ?

Reply:
Hmm-- I don't think so.  We don't want to return the exception of the stale metadata, if new metadata has been requested.